src/uct/ib/rc/accel/rc_mlx5_common.c

```diff
@@ -218,6 +218,181 @@ void uct_rc_mlx5_iface_common_tag_cleanup(uct_rc_mlx5_iface_common_t *iface,
+        desc->super.handler = (uct_rc_send_handler_t)ucs_mpool_put;
+        uct_iface_dm_mp_priv(mp)->count++;
+    }
+}
+
+static ucs_mpool_ops_t uct_dm_iface_mpool_ops = {
+    .chunk_alloc   = ucs_iface_dm_mpool_chunk_malloc,
+    .chunk_release = ucs_mpool_chunk_free,
+    .obj_init      = uct_iface_dm_mp_obj_init,
+    .obj_cleanup   = NULL
+};
+
+
+static int uct_mlx5_iface_dm_device_cmp(uct_mlx5_dm_data_t *dm_data,
+                                        uct_rc_iface_t *iface,
+                                        const uct_common_mlx5_iface_config_t *config)
+{
+    uct_ib_device_t *dev = uct_ib_iface_device(&iface->super);
+
+    return dm_data->device->ibv_context == dev->ibv_context;
+}
+
+static ucs_status_t uct_mlx5_iface_init_dm_tl(uct_mlx5_dm_data_t *data,
+                                              uct_rc_iface_t *iface,
+                                              const uct_common_mlx5_iface_config_t *config)
+{
+    ucs_status_t status;
+    struct ibv_exp_alloc_dm_attr dm_attr;
+    struct ibv_exp_reg_mr_in mr_in;
+
+    data->seg_len   = ucs_align_up(config->dm.seg_len, 8);
+    data->seg_count = ucs_align_up(config->dm.count, 8);
+    data->device    = uct_ib_iface_device(&iface->super);
+
+    dm_attr.length    = data->seg_len * data->seg_count;
+    dm_attr.comp_mask = 0;
+    data->dm          = ibv_exp_alloc_dm(data->device->ibv_context, &dm_attr);
+
+    if (data->dm == NULL) {
+        /* TODO: prompt warning? */
+        ucs_debug("ibv_exp_alloc_dm() error - On Device Memory allocation failed, %d %m", errno);
+        return UCS_OK;
+    }
+
+    memset(&mr_in, 0, sizeof(mr_in));
+    mr_in.pd           = uct_ib_iface_md(&iface->super)->pd;
+    mr_in.comp_mask    = IBV_EXP_REG_MR_DM;
+    mr_in.dm           = data->dm;
+    mr_in.length       = dm_attr.length;
+    data->mr           = ibv_exp_reg_mr(&mr_in);
+    if (data->mr == NULL) {
+        ucs_warn("ibv_exp_reg_mr() error - On Device Memory registration failed, %d %m", errno);
+        goto failed_mr;
+    } else {
+        data->start_va = ((uct_mlx5_dm_va_t*)data->dm)->start_va;
+    }
+
+    status = ucs_mpool_init(&data->mp, sizeof(uct_iface_dm_mp_priv_t),
+                            sizeof(uct_rc_iface_send_desc_t), 0, UCS_SYS_CACHE_LINE_SIZE,
+                            data->seg_count, data->seg_count,
+                            &uct_dm_iface_mpool_ops, "mlx5_dm_desc");
+    if (status != UCS_OK) {
+        goto failed_mpool;
+    }
+
+    uct_iface_dm_mp_priv(&data->mp)->dm    = data;
+    uct_iface_dm_mp_priv(&data->mp)->count = 0;
+
+complete:
+    /* DM initialization may fail due to any reason, just
+     * free resources & continue without DM */
+    return UCS_OK;
+
+failed_mpool:
+    ibv_dereg_mr(data->mr);
+failed_mr:
+    ibv_exp_free_dm(data->dm);
+    data->dm = NULL;
+    goto complete;
+}
+
+static void uct_mlx5_iface_cleanup_dm_tl(uct_mlx5_dm_data_t *data)
+{
+    if (data->dm) {
```

Comment:
maybe iface->dm could be NULL in case of no-dm, so this check would not be needed because uct_mlx_iface_dm_cleanup() already checks it

Reply:
replaced by assert